### PR TITLE
`uv tool run` suggest valid commands when command is not found

### LINKED ISSUE #4979: `uvx` should suggest valid commands when command is not found
e.g.

```
❯ uvx -q --from fastapi fastapi-cli
error: Failed to spawn: `fastapi-cli`
  Caused by: No such file or directory (os error 2)
```

Tell me what executables are available, please :)

### PR DESCRIPTION
**Pull request by blueraft**
## Summary

Resolves #4979.

## Test Plan
`cargo test`

Screenshot 2024-07-11 at 22 45 36

Screenshot 2024-07-11 at 22 45 23

### CODE REVIEW

**Comment by T-256** [comment]
> Screenshot 2024-07-11 at 21 18 05
As new user I expect to hint about `--from` when not used:
```
However, the following executables are available via `uvx --from fastapi-cli <EXECUTABLE>`:
    - fastapi.exe
```

**Comment by blueraft** [comment]
> > Screenshot 2024-07-11 at 21 18 05
> 
> As new user I expect to hint about `--from` when not used:
> 
> ```
> However, the following executables are available via `uvx --from fastapi-cli <EXECUTABLE>`:
>     - fastapi.exe
> ```

Thanks! that makes sense, I went with `uv tool run --from cmd <EXECUTABLE>` instead because `uvx` is just an alias.

File: crates/uv/src/commands/tool/run.rs [hunk 1]
```diff
@@ -118,9 +120,54 @@ pub(crate) async fn run(
...
+                    printer.stdout(),
+                    "The executable {} was not found.",
+                    command.user_display().red()
+                )?;
+                if !entry_points.is_empty() {
+                    writeln!(
+                        printer.stdout(),
+                        "However, the following executables are available: via `uv tool run --from {from} <EXECUTABLE>`",
+                    )?;
+                }
```
Comment (T-256) [inline crates/uv/src/commands/tool/run.rs:157]: Could it be ignored when user already used `--from`? I think he don't need to know about this hint. So IMO list of available executables is enough for him.
```rust
                if !entry_points.is_empty() {
                    let from_option_hint = if from == None { " via `uv tool run --from {from} <EXECUTABLE>`" } else { "" }
                    writeln!(
                        printer.stdout(),
                        "However, the following executables are available{from_option_hint}:",
                    )?;
                }
```

**Review by charliermarsh — APPROVED** [review]

**Comment by charliermarsh** [comment]
Nice, thanks!

File: crates/uv/src/commands/tool/run.rs [hunk 2]
```diff
@@ -118,9 +121,52 @@ pub(crate) async fn run(
...
+                    } else {
+                        writeln!(
+                            printer.stdout(),
+                            "The executable {} was not found.",
+                            command.to_string_lossy().red()
+                        )?;
+                        if has_from {
+                            writeln!(
+                                printer.stdout(),
+                                "However, the following executables are available:",
```
Comment (j178) [inline crates/uv/src/commands/tool/run.rs:144]: Are executables from sub-depencencies considered as *available*? For instance, `uv tool run --from fastapi`, `uvicorn` is actually an available executable.

Comment (zanieb) [reply crates/uv/src/commands/tool/run.rs:144]: I think we'll want a separate flag to opt-in to that, per https://github.com/astral-sh/uv/issues/4994

Maybe we should even warn or error if someone uses an executable that's not provided by the `--from` package?

**Pull request merged**